### PR TITLE
Introduce C API and Shared library

### PR DESCRIPTION
**Pull request by annestrand**
Introduce C API and shared library building of Clspv.
Additionally making ReflectionParser a static lib so that it's shared by both CLI and Clspv shared lib.

This PR addresses #1022

### CODE REVIEW

**Comment by rjodinchr** [comment]
Note that the C++ API does not handle errors nicely (#245). This would be the case as well for any C API.

I think we will need some testing here. That would be in the `kokoro` folder I believe.
@alan-baker do you need to do things outside of this github repository to have new tests working in github CI?

**Comment by annestrand** [comment]
Agreed - given that this just wraps around C++ API.

Thanks, will look into adding testing. Also, I somehow missed the `.clang-format` file ... d'oh!
Will fix that.

**Comment by rjodinchr** [comment]
Tests are failing because `clspv-reflection` has been renamed `clspv-reflection-cli` I believe. Is it really necessary? It would mean that this PR needs to update every test using it. But I would personally prefer not to change the name if this is possible.

**Comment by annestrand** [comment]
Ah sorry - did not account for that. Not necessary in this case - I can change it back.

I do also think there were some format errors:
```
+ ./utils/check_code_format.sh
Thu Aug 24 16:00:23 UTC 2023: Check formatting...
Found formatting errors!
```

**Review by rjodinchr — COMMENTED** [review]
I feel like we should not create an `api` folder for it.
We could merge every part into the existing structure:
- `api/CMakeLists.txt` merge in `lib/CMakeLists.txt`
- `api/Clspv.cpp` merge in `lib/Compiler.cpp`
- `api/Clspv.h` merge in `include/clspv/Compiler.h`
- `api/README.md` moves to `docs`, with a mention in the top README, and maybe mention of the C++ API.

@alan-baker will have the final word on that, so wait for his point-of-view before changing everything.

File: api/Clspv.h
Comment (annestrand) [reply api/Clspv.h:85]: Done 👍

File: tools/reflection/main.cpp
Comment (annestrand) [reply tools/reflection/main.cpp:63]: Got it - I went ahead and dropped the commit for this.

File: tools/reflection/ReflectionParser.cpp
Comment (alan-baker) [inline tools/reflection/ReflectionParser.cpp:333]: This tool is deprecated and should not be used. APIs should parse the reflection data into their own internal data structures.

Please revert changes to this directory.

File: api/Clspv.cpp
Comment (alan-baker) [inline api/Clspv.cpp:136]: This is not the right function to use. It has intentionally not been kept up-to-date with interface changes. Users of clspv are responsible for parsing the embedded reflection data in the SPIR-V.

File: api/README.md
Comment (alan-baker) [inline api/README.md:1]: This belongs in docs as `C_API.md` or a similar name.

File: CMakeLists.txt
Comment (alan-baker) [inline CMakeLists.txt:191]: I agree with Romaric that a new directory is not desirable.

File: api/Clspv.cpp
Comment (annestrand) [reply api/Clspv.cpp:29]: It's technically different in that we now have this container/context object that wraps the C++ items. 
Hence why thread safety should be considered here.

Did you mean user/app side here? If so, yes we could explicitly tell users (say in the C_API.md doc) that they are responsible for synchronization here. I prefer the existing approach, but I'm open to your suggestion if we mention this in docs.

File: api/Clspv.h
Comment (annestrand) [reply api/Clspv.h:44]: My oversight there perhaps 😅 - Will update and push.

File: tools/reflection/ReflectionParser.cpp
Comment (annestrand) [reply tools/reflection/ReflectionParser.cpp:333]: My mistake - I did not know this tool was deprecated. Will update and push.

File: api/Clspv.cpp
Comment (annestrand) [reply api/Clspv.cpp:136]: Ack.

File: api/README.md
Comment (annestrand) [reply api/README.md:1]: Ack.

File: CMakeLists.txt
Comment (annestrand) [reply CMakeLists.txt:191]: Ack.

File: api/Clspv.h
Comment (annestrand) [reply api/Clspv.h:44]: Actually now I remember... we need to pass container ptr by reference (otherwise its just a copy).

We could modify create function to instead return ptr to created container. That might be a better approach.

**Review by rjodinchr — COMMENTED** [review]
`ClspvContainer` is not needed anymore. It made sense when you had to pass data between the compiler function and the reflection parser one. But now it is not used as you only have the compiler function.

Removing it will largely reduce the size of this PR. Then you might consider moving `include/clspv/Clspv.h` in `include/clspv/Compiler.h`, and `lib/Clspv.cpp` in `lib/Compiler.cpp`

**Comment by annestrand** [comment]
@rjodinchr - Sure, I can merge the separate header and source to `Compiler.h/.cpp` respectively.

However, on removing the container object, I don't think there is a clean way to free a dyn. alloced std::vector or std::string from the user side here when they/user no longer needs either the spv code or the build log respectively (since `clspv::CompileFromSourcesString` requires these C++ objs).

I think the merging to `Compiler.h/.cpp` should help reduce the PR size here.

**Comment by rjodinchr** [comment]
> However, on removing the container object, I don't think there is a clean way to free a dyn. alloced std::vector or std::string from the user side here when they/user no longer needs either the spv code or the build log respectively (since `clspv::CompileFromSourcesString` requires these C++ objs).

This C-API is already coying `programs` (the C input) to `vprograms` (C++ `std::vector<std::string>`). I think it would need to do something similar for the outputs: meaning to copy the `std::vector`/`std::string` into buffers allocated with `malloc` that would be the user's responsibility to free.

We may provide a helper function to free those objects (maybe in the header so that the content is clearly visible?). But I think it would be important to have both versions in the doc (manual free, and using the helper function).

**Comment by annestrand** [comment]
That's a fair point. Now, if that same `clspv::CompileFromSourcesString` had an overload at some point to allow passing C type strings, then we would not copy at all in the container approach I believe. Benefit of this overload could be if clspv is used inside of an OpenCL driver impl. where maintainer would not need to copy input strings to a C++ container.

That's out of scope for this PR, so back to your suggestion - I'm okay with removal of the container here and go with your approach. My only heartburn was the extra memcpy for the outputs, but I think I can look past that here from your point above.

File: docs/C_API.md
Comment (rjodinchr) [inline docs/C_API.md:29]: This is the default value, thus not really needed

File: include/clspv/Compiler.h
Comment (rjodinchr) [inline include/clspv/Compiler.h:90]: free of nullptr is allowed, we should not check for it:
> If ptr is a null pointer, the function does nothing.
https://en.cppreference.com/w/c/memory/free

File: lib/Compiler.cpp
Comment (rjodinchr) [inline lib/Compiler.cpp:665]: This should not be part of this PR

File: include/clspv/Compiler.h
Comment (annestrand) [reply include/clspv/Compiler.h:90]: Ack

File: docs/C_API.md
Comment (annestrand) [reply docs/C_API.md:29]: Ack

**Review by rjodinchr — APPROVED** [review]

**Comment by annestrand** [comment]
Hmm, not sure what went wrong with `CI-ubuntu-gcc-debug` run? Tried clicking "Details" to get log but I get a DNS error.

**Comment by alan-baker** [comment]
I've restarted that build. It was an internal infrastructure error so it couldn't publish the logs.

**Review by alan-baker — APPROVED** [review]

**Comment by rjodinchr** [comment]
Here what I found in the logs:
```
[2192/2195] Linking CXX executable bin/clspv
FAILED: bin/clspv 
: && /usr/bin/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -fno-rtti -fno-exceptions -Werror=unused-variable -Werror=switch -g  tools/driver/CMakeFiles/clspv.dir/main.cpp.o -o bin/clspv  lib/libclspv_core.a  third_party/llvm/lib/libclangCodeGen.a  third_party/llvm/lib/libLLVMCoverage.a  third_party/llvm/lib/libLLVMLTO.a  third_party/llvm/lib/libLLVMPasses.a  third_party/llvm/lib/libLLVMCodeGen.a  third_party/llvm/lib/libLLVMObjCARCOpts.a  third_party/llvm/lib/libLLVMCoroutines.a  third_party/llvm/lib/libLLVMIRPrinter.a  third_party/llvm/lib/libLLVMipo.a  third_party/llvm/lib/libLLVMLinker.a  third_party/llvm/lib/libLLVMInstrumentation.a  third_party/llvm/lib/libLLVMVectorize.a  third_party/llvm/lib/libLLVMBitWriter.a  third_party/llvm/lib/libLLVMCodeGenTypes.a  third_party/llvm/lib/libLLVMTarget.a  third_party/llvm/lib/libLLVMExtensions.a  third_party/llvm/lib/libclangFrontend.a  third_party/llvm/lib/libclangDriver.a  third_party/llvm/lib/libLLVMWindowsDriver.a  third_party/llvm/lib/libclangParse.a  third_party/llvm/lib/libLLVMOption.a  third_party/llvm/lib/libclangSerialization.a  third_party/llvm/lib/libclangSema.a  third_party/llvm/lib/libclangAnalysis.a  third_party/llvm/lib/libclangASTMatchers.a  third_party/llvm/lib/libLLVMFrontendHLSL.a  third_party/llvm/lib/libclangEdit.a  third_party/llvm/lib/libclangAST.a  third_party/llvm/lib/libLLVMFrontendOpenMP.a  third_party/llvm/lib/libLLVMScalarOpts.a  third_party/llvm/lib/libLLVMAggressiveInstCombine.a  third_party/llvm/lib/libLLVMInstCombine.a  third_party/llvm/lib/libLLVMTransformUtils.a  third_party/llvm/lib/libLLVMAnalysis.a  third_party/llvm/lib/libLLVMProfileData.a  third_party/llvm/lib/libLLVMSymbolize.a  third_party/llvm/lib/libLLVMDebugInfoPDB.a  third_party/llvm/lib/libLLVMDebugInfoMSF.a  third_party/llvm/lib/libLLVMDebugInfoBTF.a  third_party/llvm/lib/libLLVMDebugInfoDWARF.a  third_party/llvm/lib/libLLVMObject.a  third_party/llvm/lib/libLLVMIRReader.a  third_party/llvm/lib/libLLVMAsmParser.a  third_party/llvm/lib/libLLVMMCParser.a  third_party/llvm/lib/libLLVMTextAPI.a  third_party/llvm/lib/libLLVMBitReader.a  third_party/llvm/lib/libLLVMCore.a  third_party/llvm/lib/libLLVMRemarks.a  third_party/llvm/lib/libLLVMMC.a  third_party/llvm/lib/libLLVMBinaryFormat.a  third_party/llvm/lib/libLLVMDebugInfoCodeView.a  third_party/llvm/lib/libclangLex.a  third_party/llvm/lib/libclangBasic.a  third_party/llvm/lib/libLLVMTargetParser.a  third_party/llvm/lib/libclangSupport.a  third_party/llvm/lib/libLLVMBitstreamReader.a  third_party/llvm/lib/libLLVMSupport.a  third_party/llvm/lib/libLLVMDemangle.a  -lrt  -ldl  -lpthread  -lm  /usr/lib/x86_64-linux-gnu/libz.so && :
/usr/bin/ld: failed to set dynamic section sizes: Memory exhausted
collect2: error: ld returned 1 exit status
```

**Comment by annestrand** [comment]
Yep, looks like shared lib building is adding some mem strain to that build machine (for GCC debug at least).

Do we know what the specs are for that machine/container? I would like to try reproducing it locally (I might have a workaround).

**Comment by annestrand** [comment]
Okay, looks like GCC debug is clean. Now I'm getting new CI issue here - `CI-clvk-ubuntu-clang-debug`:
```
-- Using SPIR-V Headers from
          /tmpfs/src/github/clvk/external/SPIRV-Headers
CMake Error at external/SPIRV-LLVM-Translator/CMakeLists.txt:82 (find_package):
  Could not find a configuration file for package "LLVM" that is compatible
  with requested version "17.0.0".

  The following configuration files were considered but not accepted:

    /tmpfs/src/github/clvk/build/external/clspv/third_party/llvm/lib/cmake/llvm/LLVMConfig.cmake, version: 18.0.0git
    /bin/clang-8.0.0/lib/cmake/llvm/LLVMConfig.cmake, version: 8.0.0
```
@rjodinchr - Any clues for this one?

**Comment by alan-baker** [comment]
> Okay, looks like GCC debug is clean. Now I'm getting new CI issue here - `CI-clvk-ubuntu-clang-debug`:
> 
> ```
> -- Using SPIR-V Headers from
>           /tmpfs/src/github/clvk/external/SPIRV-Headers
> CMake Error at external/SPIRV-LLVM-Translator/CMakeLists.txt:82 (find_package):
>   Could not find a configuration file for package "LLVM" that is compatible
>   with requested version "17.0.0".
> 
>   The following configuration files were considered but not accepted:
> 
>     /tmpfs/src/github/clvk/build/external/clspv/third_party/llvm/lib/cmake/llvm/LLVMConfig.cmake, version: 18.0.0git
>     /bin/clang-8.0.0/lib/cmake/llvm/LLVMConfig.cmake, version: 8.0.0
> ```
> 
> @rjodinchr - Any clues for this one?

That's a problem with LLVM-SPIRV-Translator not being up to date with the latest LLVM. Not your problem.

**Pull request merged**
